packages/react-utils/src/components/SearchBar/index.tsx

```diff
@@ -0,0 +1,78 @@
+import React, { ChangeEvent } from 'react';
+import { SearchOutlined } from '@ant-design/icons';
+import { Input, Space } from 'antd';
+import reducerRegistry from '@onaio/redux-reducer-registry';
+import {
+  reducerName,
+  filteredDataReducer,
+  fetchFilteredDataAction,
+  removeFilteredDataAction,
+} from '../../ducks';
+import { useDispatch } from 'react-redux';
+import { SizeType } from 'antd/lib/config-provider/SizeContext';
+import { isEqual } from 'lodash';
+
+reducerRegistry.register(reducerName, filteredDataReducer);
+
+/** SearchBar props interface  */
+export interface SearchBarDefaultProps {
+  placeholder?: string;
+  size?: SizeType;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  data: any[];
+}
+
+/** function type for custom onChangeHandler functions */
+export type OnChangeType = (event: ChangeEvent<HTMLInputElement>) => void;
+
+/** search bar props interface */
+export interface SearchBarProps extends SearchBarDefaultProps {
+  onChangeHandler?: OnChangeType;
+}
+
+const defaultProps: SearchBarProps = {
+  placeholder: '',
+  size: 'large',
+  data: [],
+};
+
+const SearchBar = (props: SearchBarProps) => {
+  const { placeholder, size, data, onChangeHandler } = props;
+  const dispatch = useDispatch();
+
+  const customOnChangeHandler = (e: ChangeEvent<HTMLInputElement>) => {
+    const currentValue = e.target.value;
+    const filteredData = [];
+    for (const datum of data) {
+      const values = Object.values(datum);
+      for (const val of values) {
+        if ((val as string).toLowerCase().includes(currentValue.toLowerCase())) {
+          filteredData.push(datum);
+        }
+      }
```

Comment:
I kind of feel iffy about this structure due to :
- looks like there is an implied assumption of what data would look like, reducing the robustness of the component, 
- also these could probably match against alot of data objects, and would require the user do full text matches for accurate results, 
- could be computationally expensive

I propose we sync the filter state with the url, such that to each component can parse the url, extract the filters and pass them on to a selector that knows how to apply the filters to the required data.
This way we get a few benefits:
- given reducer know how best to filter data, what fields to match and which need not be matched,
- ideally, selector used should be memoized, thus giving us a speed bonus
- we can share urls that have filter context encoded in them
- it can allow for some more robust patterns like having more than one search component on a page.

Reply:
@KipSigei let me know what you think?

Reply:
@p-netm, In that case, we can borrow the implementation in Reveal right? Basically the `Search` should debounce the users input, push it the value to the URL and the component making use of the search will get the value and use selectors to filter the data?

Reply:
@kelvin-muchiri yeah, pretty much.